src/sentry/api/authentication.py

```diff
@@ -298,6 +299,55 @@ def authenticate(self, request: Request):
 class UserAuthTokenAuthentication(StandardAuthentication):
     token_name = b"bearer"
 
+    def _find_or_update_token_by_hash(self, token_str: str) -> ApiToken | ApiTokenReplica:
+        """
+        Find token by hash or update token's hash value if only found via plaintext.
+
+        1. Hash provided plaintext token.
+        2. Perform lookup based on hashed value.
+        3. If found, return the token.
+        4. If not found, search for the token based on its plaintext value.
+        5. If found, update the token's hashed value and return the token.
+        6. If not found via hash or plaintext value, raise AuthenticationFailed
+
+        Returns `ApiTokenReplica` if running in REGION silo or
+        `ApiToken` if running in CONTROL silo.
+        """
+
+        hashed_token = hashlib.sha256(token_str.encode()).hexdigest()
+
+        if SiloMode.get_current_mode() == SiloMode.REGION:
+            try:
+                # Try to find the token by its hashed value first
+                return ApiTokenReplica.objects.get(hashed_token=hashed_token)
```

Comment:
Do we have an index on this column? We could fall back to a table scan here.

Reply:
https://github.com/getsentry/sentry/pull/66947

Good idea :)